src/mbgl/style/expression/number_format.cpp

```diff
@@ -0,0 +1,276 @@
+                fractPart = round(fractPart * pow(10, maxPrecision));
+                fractPart = fractPart / pow(10, maxPrecision);
+                fractPartStr = std::to_string(fractPart).substr(2, std::string::npos);
+                fractPartStr.resize(maxPrecision);
+
+                // Increment the integer part if the fractional part rounds up to 1
+                if (fractPart >= 1) {
+                    integerPart++;
+                }
+            }
+            fractPartStr.insert(0, ".");
+        } else {
+            // Round the value to the nearest integer in case no decimals are requested
+            integerPart = round(evaluatedNumber);
+        }
+        ss << integerPart << fractPartStr;
+    }
+
+    return ss.str();
+}
+
+void NumberFormat::eachChild(const std::function<void(const Expression&)>& visit) const {
+    visit(*number);
+    if (locale) visit(**locale);
+    if (currency) visit(**currency);
+    if (minFractionDigits) visit(**minFractionDigits);
+    if (maxFractionDigits) visit(**maxFractionDigits);
+}
+
+bool NumberFormat::operator==(const Expression& e) const {
+    if (e.getKind() == Kind::NumberFormat) {
+        auto rhs = static_cast<const NumberFormat*>(&e);
+        if ((locale && (!rhs->locale || **locale != **rhs->locale)) ||
+            (!locale && rhs->locale)) {
+            return false;
+        }
+        if ((currency && (!rhs->currency || **currency != **rhs->currency)) ||
+            (!currency && rhs->currency)) {
+            return false;
+        }
+        if ((minFractionDigits && (!rhs->minFractionDigits || **minFractionDigits != **rhs->minFractionDigits)) ||
+            (!minFractionDigits && rhs->minFractionDigits)) {
+            return false;
+        }
+        if ((maxFractionDigits && (!rhs->maxFractionDigits || **maxFractionDigits != **rhs->maxFractionDigits)) ||
+            (!maxFractionDigits && rhs->maxFractionDigits)) {
+            return false;
+        }
+        return *number == *rhs->number;
+    }
+    return false;
+}
+
+std::vector<optional<Value>> NumberFormat::possibleOutputs() const {
+    return { nullopt };
+}
+
+using namespace mbgl::style::conversion;
+ParseResult NumberFormat::parse(const Convertible& value, ParsingContext& ctx) {
+    std::size_t length = arrayLength(value);
+
+    if (length != 3) {
+        ctx.error("Expected two arguments, but found " + util::toString(length) + " instead.");
+        return ParseResult();
+    }
+
+    ParseResult number = ctx.parse(arrayMember(value, 1), 1, {type::Number});
+    if (!number) {
+        ctx.error("Failed to parse the number.");
+        return ParseResult();
+    }
+
+    type::Type type = (*number)->getType();
+    if (!type.is<type::NumberType>()) {
+        ctx.error("Expected argument of type number, but found " + toString(type) + " instead.");
+        return ParseResult();
+    }
+
+    auto options = arrayMember(value, 2);
+    if (!isObject(options)) {
```

Comment:
A good unit test would be passing an empty options, which apparently is allowed by the spec. In this case I think the number returned should be the same as the original unformatted number.